### PR TITLE
cli: support for m3 namespaces

### PR DESCRIPTION
**Pull request by mts-aiven**
Add support for M3 namespaces.

Added commands: avn service m3 namespace {add, delete, list}

### CODE REVIEW

**Comment by mts-aiven** [comment]
```
(aiven-client) [mts@localhost aiven-client]$ scripts/avn --url https://aiven-rest-aiven-public-mts-test.avns.net service create -t m3db m3db-kekkonen --cloud google-europe-west1 -p business-8
(aiven-client) [mts@localhost aiven-client]$ scripts/avn --url https://aiven-rest-aiven-public-mts-test.avns.net service m3 namespace add m3db-kekkonen tunkati --ns-type aggregated --ns-ret 8d --ns-res 1d
Namespace(config='/home/mts/.config/aiven/aiven-client.json', command='service', auth_ca=None, auth_token=None, show_http=False, url='https://aiven-rest-aiven-public-mts-test.avns.net', project=None, name='m3db-kekkonen', ns_name='tunkati', ns_type='aggregated', ns_ret='8d', ns_res='1d', func=<bound method AivenCLI.service__m3__namespace__add of <aiven.client.cli.AivenCLI object at 0x7fbcd928e280>>)
(aiven-client) [mts@localhost aiven-client]$ scripts/avn --url https://aiven-rest-aiven-public-mts-test.avns.net service m3 namespace list m3db-kekkonen 
NAME     TYPE          RESOLUTION  OPTIONS.RETENTION_OPTIONS.RETENTION_PERIOD_DURATION  OPTIONS.RETENTION_OPTIONS.BLOCKSIZE_DURATION
=======  ============  ==========  ===================================================  ============================================
default  unaggregated                                                                   2h
tunkati  aggregated    1d          8d                                                   4h
(aiven-client) [mts@localhost aiven-client]$ scripts/avn --url https://aiven-rest-aiven-public-mts-test.avns.net service m3 namespace list m3db-kekkonen --json
[
    {
        "name": "default",
        "options": {
            "retention_options": {
                "blocksize_duration": "2h"
            }
        },
        "type": "unaggregated"
    },
    {
        "name": "tunkati",
        "options": {
            "retention_options": {
                "blocksize_duration": "4h",
                "retention_period_duration": "8d"
            }
        },
        "resolution": "1d",
        "type": "aggregated"
    }
]
(aiven-client) [mts@localhost aiven-client]$ scripts/avn --url https://aiven-rest-aiven-public-mts-test.avns.net service m3 namespace delete m3db-kekkonen tunkati

aiven-client) [mts@localhost aiven-client]$ scripts/avn --url https://aiven-rest-aiven-public-mts-test.avns.net service m3 namespace -h
usage: avn service m3 namespace [-h] {add,delete,list} ...

positional arguments:
  {add,delete,list}
    add              List M3 namespaces
    delete           List M3 namespaces
    list             List M3 namespaces

optional arguments:
  -h, --help         show this help message and exit

```

File: aiven/client/cli.py [hunk 1]
```diff
@@ -1684,6 +1684,56 @@ def service__index_delete(self):
...
+
+    @arg.project
+    @arg.service_name
+    @arg.ns_name
+    @arg.ns_type
+    @arg.ns_ret
+    @arg.ns_res
+    def service__m3__namespace__add(self):
+        """Add M3 namespaces"""
+        print(self.args)
```
Comment (tsharju) [inline aiven/client/cli.py:1727]: Maybe not print the arguments here. Or rather something bit prettier if we want to show output to the user.

File: aiven/client/client.py [hunk 2]
```diff
@@ -400,6 +400,47 @@ def get_service_topic(self, project, service, topic):
...
+    def list_m3_namespaces(self, project, service):
+        service_res = self.get_service(project=project, service=service)
+        return service_res.get("user_config", {}).get("namespaces", [])
+
+    def delete_m3_namespace(self, project, service, ns_name):
+        service_res = self.get_service(project=project, service=service)
+        old_namespaces = service_res.get("user_config", {}).get("namespaces", [])
+        new_namespaces = [ns for ns in old_namespaces if ns["name"] != ns_name]
+        if len(old_namespaces) == len(new_namespaces):
+            raise KeyError(f"Namespace '{ns_name}' does not exist")
```
Comment (tsharju) [inline aiven/client/client.py:412]: Should rather print a meaningful error than raising exception which is then visible to the users as a stacktrace.

File: aiven/client/client.py [hunk 3]
```diff
@@ -400,6 +400,47 @@ def get_service_topic(self, project, service, topic):
...
+        return self.verify(
+            self.put,
+            self.build_path("project", project, "service", service),
+            body={"user_config": {
+                "namespaces": new_namespaces
+            }},
+        )
+
+    def add_m3_namespace(self, project, service, ns_name, ns_type, ns_ret, ns_res=None):
```
Comment (tsharju) [inline aiven/client/client.py:421]: Maybe we could check some of the values for validity here. For example, the namespace type (aggregated/unaggregated) seems to be possible to accidentally type wrong and it is stored to the user config as the API does not seem to validate it yet at least.

File: aiven/client/cliarg.py [hunk 4]
```diff
@@ -93,6 +93,10 @@ def wrapped(self):
 arg.retention_bytes = arg("--retention-bytes", type=int, help="Retention limit in bytes (default: unlimited)")
 arg.service_name = arg("name", help="Service name")
 arg.service_type = arg("-t", "--service-type", help="Type of service (see 'service list-types')")
+arg.ns_name = arg("ns_name", help="Namespace name")
+arg.ns_type = arg("--ns-type", help="Namespace type ('aggregated' or 'unaggregated')", required=True)
+arg.ns_ret = arg("--ns-ret", help="Namespace retention period", required=True)
+arg.ns_res = arg("--ns-res", help="Namespace resolution")
```
Comment (tsharju) [inline aiven/client/cliarg.py:99]: nit: perhaps `retention` and `resolution` instead of the abbreviated argument names would be a bit more readable?

**Review by tsharju — DISMISSED** [review]

File: aiven/client/cli.py (hunk 1 above)
Comment (mts-aiven) [reply aiven/client/cli.py]: left my debug print there by mistake

File: aiven/client/cliarg.py (hunk 4 above)
Comment (mts-aiven) [reply aiven/client/cliarg.py]: ok, changed

File: aiven/client/client.py (hunk 2 above)
Comment (mts-aiven) [reply aiven/client/client.py:412]: Fixed in cli.py by catching this and raising args.UserError, which seems to print out nicely. This seems to be done in cli.py for other errors.

File: aiven/client/client.py (hunk 3 above)
Comment (mts-aiven) [reply aiven/client/client.py:421]: Could add here. I think it would be more critical for the API to validate incoming stuff though.

Comment (mts-aiven) [reply aiven/client/client.py:422]: added validations

**Comment by mts-aiven** [comment]
```
(aiven-client) ~/git/aiven-client:mts-m3-namespaces$ scripts/avn --url https://aiven-rest-aiven-public-mts-test.avns.net service m3 namespace
usage: avn service m3 namespace [-h] {add,delete,list} ...

positional arguments:
  {add,delete,list}
    add              Add M3 namespaces
    delete           Delete M3 namespaces
    list             List M3 namespaces

optional arguments:
  -h, --help         show this help message and exit

(aiven-client) ~/git/aiven-client:mts-m3-namespaces$ scripts/avn --url https://aiven-rest-aiven-public-mts-test.avns.net service m3 namespace list m3db-kekkonen
NAME     TYPE          RESOLUTION  OPTIONS.RETENTION_OPTIONS.RETENTION_PERIOD_DURATION  OPTIONS.RETENTION_OPTIONS.BLOCKSIZE_DURATION
=======  ============  ==========  ===================================================  ============================================
default  unaggregated                                                                   2h
tff      aggregated    1d          8d                                                   4h
huuhaa   aggregated    1d          8d                                                   4h

(aiven-client) ~/git/aiven-client:mts-m3-namespaces$ scripts/avn --url https://aiven-rest-aiven-public-mts-test.avns.net service m3 namespace delete m3db-kekkonen huuhaa
(aiven-client) ~/git/aiven-client:mts-m3-namespaces$ scripts/avn --url https://aiven-rest-aiven-public-mts-test.avns.net service m3 namespace add -h
usage: avn service m3 namespace add [-h] [--project PROJECT] --ns-type NS_TYPE --ns-retention NS_RETENTION [--ns-resolution NS_RESOLUTION] name ns_name

positional arguments:
  name                  Service name
  ns_name               Namespace name

optional arguments:
  -h, --help            show this help message and exit
  --project PROJECT     Project name to use, default None
  --ns-type NS_TYPE     Namespace type ('aggregated' or 'unaggregated')
  --ns-retention NS_RETENTION
                        Namespace retention period
  --ns-resolution NS_RESOLUTION
                        Namespace resolution

(aiven-client) ~/git/aiven-client:mts-m3-namespaces$ scripts/avn --url https://aiven-rest-aiven-public-mts-test.avns.net service m3 namespace add m3db-kekkonen ggg --ns-type aggregated --ns-retention 7d --ns-resolution 10m

(aiven-client) ~/git/aiven-client:mts-m3-namespaces$ scripts/avn --url https://aiven-rest-aiven-public-mts-test.avns.net service m3 namespace list m3db-kekkonen
NAME     TYPE          RESOLUTION  OPTIONS.RETENTION_OPTIONS.RETENTION_PERIOD_DURATION  OPTIONS.RETENTION_OPTIONS.BLOCKSIZE_DURATION
=======  ============  ==========  ===================================================  ============================================
default  unaggregated                                                                   2h
tff      aggregated    1d          8d                                                   4h
ggg      aggregated    10m         7d                                                   2h




```

**Comment by fingon** [comment]
Some sort of namespace update might be needed for unaggregated namespace, as REST endpoint will reject its delete (there has to be always 1 configured) or add (-//-).

**Comment by mts-aiven** [comment]
Added support for more fields and update command.

```
(aiven-client) ~/git/aiven-client:mts-m3-namespaces$ scripts/avn --url https://aiven-rest-aiven-public-mts-test.avns.net service m3 namespace update m3db-kekkonen sekava --ns-retention 10m --ns-resolution 10s --ns-blocksize-dur 20s --ns-block-data-expiry-dur 20s --ns-buffer-future-dur 10s --ns-buffer-past-dur 10s --ns-writes-to-commitlog true
(aiven-client) ~/git/aiven-client:mts-m3-namespaces$ scripts/avn --url https://aiven-rest-aiven-public-mts-test.avns.net service m3 namespace list m3db-kekkonen 
NAME      TYPE          RESOLUTION  RETENTION_PERIOD_DURATION  BLOCKSIZE_DURATION  BLOCK_DATA_EXPIRY_DURATION  BUFFER_FUTURE_DURATION  BUFFER_PAST_DURATION  WRITES_TO_COMMITLOG
========  ============  ==========  =========================  ==================  ==========================  ======================  ====================  ===================
default   unaggregated                                         2h
tff       aggregated    1d          8d                         4h
humppati  aggregated    10s         40s                        10s                 40s                         30s                     20s                   true
sekava    aggregated    10s         10m                        20s                 20s                         10s                     10s                   true

```

**Comment by mts-aiven** [comment]
Missing removing set values from configuration (to use default values), could do that in follow up pull if/when required.

**Review by tsharju — APPROVED** [review]

**Pull request merged**